### PR TITLE
Add libunrar

### PR DESCRIPTION
**Pull request by sshockwave**
The name `unrar` is unfortunately taken by the [python package](https://pypi.org/project/unrar/) ([feedstock](https://github.com/conda-forge/unrar-feedstock)), which turns out to be non-functioning unless this C++ library is installed.

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by sshockwave** [comment]
The library build has `-DSILENT` so the utility `unrar` can't be dynamically linked to the library. I'm leaving the option as-is to ensure consistency with the original behaviors.

**Comment by sshockwave** [comment]
@conda-forge/help-c-cpp, ready for review!

File: recipes/libunrar/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+source:
+  url: https://www.rarlab.com/rar/unrarsrc-{{ version }}.tar.gz
+  sha256: 120936e41f826cd63d77a580aeea64c1b79fd3e2434f58ce8184e783b51e5b01
+  patches:
+    - 0001-REMOVE-HARDCODED-MAKEFILE-ENV.patch  # [unix]
+    - 0002-REMOVE-INSTALL-OPTION.patch  # [osx]
+    - 0003-CHANGE-DYLIB-EXTENSION.patch  # [osx]
+
+build:
```
Comment (xhochy) [inline recipes/libunrar/meta.yaml:16]: How is the ABI stability of this lib? It will need a matching run_exports

**Review by xhochy — CHANGES_REQUESTED** [review]

File: recipes/libunrar/meta.yaml (hunk 1 above)
Comment (sshockwave) [reply recipes/libunrar/meta.yaml:16]: The ABI stability is really, really bad and changes on every minor version. Not even its own utility can link to the lib. I think it's only meant to be used statically.

That said, this is the official implementation of RAR unpacking and the community alternatives are not able to achieve feature parity with the newest non-free format.

File: recipes/libunrar/build.sh [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+make unrar
+mkdir -p "${PREFIX}/bin"
+DESTDIR="${PREFIX}" make install-unrar
+
+mkdir -p "${PREFIX}/include/unrar"
+cp *.hpp "${PREFIX}/include/unrar"
+
+# CFEP-18
+rm "${PREFIX}/lib/libunrar.a"
```
Comment (carterbox) [inline recipes/libunrar/build.sh:25]: 👍🏼

File: recipes/libunrar/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  run_exports:
+    - {{ pin_subpackage('libunrar', exact=True) }}
```
Comment (carterbox) [inline recipes/libunrar/meta.yaml:18]:
```suggestion
    - {{ pin_subpackage('libunrar', max_pin='x.x.x') }}
```
Exact means including the build string. You said the ABI changes with every patch version.

File: recipes/libunrar/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  build:
+    - {{ compiler('cxx') }}
+    - make  # [unix]
+
+test:
+  commands:
+    - unrar
+
+    # CFEP-18
+    - test ! -f ${PREFIX}/lib/libunrar.a  # [unix]
```
Comment (carterbox) [inline recipes/libunrar/meta.yaml:31]:
```suggestion
test:
  commands:
    - unrar
    - test -f $PREFIX/lib/libunrar.so                 # [linux]
    - test -f $PREFIX/lib/libunrar.dylib              # [osx]
    - if not exist %LIBRARY_LIB%\libunrar.lib exit 1  # [win]
    - if not exist %LIBRARY_BIN%\libunrar.dll exit 1  # [win]
    - test -f $PREFIX/include/unrar/headers.hpp  # [unix]
    - if not exist %LIBRARY_INC%\unrar\headers.hpp exit 1  # [win]
    # CFEP-18
    - test -f ${PREFIX}/lib/libunrar.${DYLIB}  # [unix]
```

**Review by carterbox — CHANGES_REQUESTED** [review]
@sshockwave, could you please open a PR over at @conda-forge/unrar to add yourself as a contributor and add this package as a runtime dependency.

File: recipes/libunrar/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  commands:
+    - unrar
+    - test -f $PREFIX/lib/libunrar.so                 # [linux]
+    - test -f $PREFIX/lib/libunrar.dylib              # [osx]
+    - if not exist %LIBRARY_LIB%\UnRAR.lib exit 1  # [win]
+    - if not exist %LIBRARY_BIN%\unrar.dll exit 1  # [win]
+    - test -f $PREFIX/include/unrar/headers.hpp  # [unix]
+    - if not exist %LIBRARY_INC%\unrar\headers.hpp exit 1  # [win]
+    # CFEP-18
+    - test -f ${PREFIX}/lib/libunrar.${DYLIB}  # [unix]
```
Comment (carterbox) [inline recipes/libunrar/meta.yaml:36]:
```suggestion
    - test ! -f ${PREFIX}/lib/libunrar.a  # [unix]
```

File: recipes/libunrar/meta.yaml (hunk 3 above)
Comment (sshockwave) [reply recipes/libunrar/meta.yaml]: I see. Thanks!

**Comment by sshockwave** [comment]
> @sshockwave, could you please open a PR over at @conda-forge/unrar to add yourself as a contributor and add this package as a runtime dependency.

Sure. Will do!

**Pull request merged**